rust/lance-encoding/src/encodings/physical/dictionary.rs

```diff
@@ -0,0 +1,434 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::sync::Arc;
+
+use arrow_array::types::UInt64Type;
+use arrow_array::{Array, ArrayRef, DictionaryArray, StringArray, UInt64Array};
+use futures::{future::BoxFuture, FutureExt};
+
+use crate::{
+    decoder::{PageScheduler, PrimitivePageDecoder},
+    encoder::{ArrayEncoder, EncodedArray},
+    format::pb,
+    EncodingsIo,
+};
+
+use crate::decoder::LogicalPageDecoder;
+use crate::encodings::logical::primitive::PrimitiveFieldDecoder;
+
+use arrow_schema::DataType;
+use bytes::BytesMut;
+use lance_core::Result;
+use std::collections::HashMap;
+
+use crate::encodings::utils::new_primitive_array;
+use arrow_array::cast::AsArray;
+
+#[derive(Debug)]
+pub struct DictionaryPageScheduler {
+    indices_scheduler: Arc<dyn PageScheduler>,
+    items_scheduler: Arc<dyn PageScheduler>,
+    size: u64,
+}
+
+impl DictionaryPageScheduler {
+    pub fn new(
+        indices_scheduler: Arc<dyn PageScheduler>,
+        items_scheduler: Arc<dyn PageScheduler>,
+        size: u64,
+    ) -> Self {
+        Self {
+            indices_scheduler,
+            items_scheduler,
+            size,
+        }
+    }
+}
+
+impl PageScheduler for DictionaryPageScheduler {
+    fn schedule_ranges(
+        &self,
+        ranges: &[std::ops::Range<u64>],
+        scheduler: &Arc<dyn EncodingsIo>,
+        top_level_row: u64,
+    ) -> BoxFuture<'static, Result<Box<dyn PrimitivePageDecoder>>> {
+        // We want to decode indices and items
+        // e.g. indices [0, 1, 2, 0, 1, 0]
+        // items (dictionary) ["abcd", "hello", "apple"]
+        // This will map to ["abcd", "hello", "apple", "abcd", "hello", "abcd"]
+        // We decode all the items during scheduling itself
+        // These are used to rebuild the string later
+
+        // Schedule indices for decoding
+        let indices_page_decoder =
+            self.indices_scheduler
+                .schedule_ranges(ranges, scheduler, top_level_row);
+
+        // Schedule items for decoding
+        let items_page_decoder =
+            self.items_scheduler
+                .schedule_ranges(&[0..self.size], scheduler, top_level_row);
+
+        let copy_size = self.size;
+
+        async move {
+            let items_decoder: Arc<dyn PrimitivePageDecoder> = Arc::from(items_page_decoder.await?);
+
+            let mut primitive_wrapper = PrimitiveFieldDecoder::new_from_data(
+                items_decoder.clone(),
+                DataType::Utf8,
+                copy_size,
+            );
+
+            // Decode all items
+            let drained_task = primitive_wrapper.drain(copy_size)?;
+            let items_decode_task = drained_task.task;
+            let decoded_dict = items_decode_task.decode()?;
```

Comment:
It's not ideal we do the decoding here but I think the correct fix is probably better suited for a future PR.  So let's leave it as is for now.  Ideally, all potentially CPU intensive decoding work happens in the decode task but it's a bit challenging with dictionary because multiple decode tasks will share the same dictionary item.  We could use a mutex though and let the first decode task involving the dictionary do the actual decoding.  Also, in the future, we might want to decode in some kind of initialization step and cache the results for multiple scans.